flake.nix

```diff
@@ -21,35 +21,13 @@
 
         nixos = import ./nixos/lib { lib = final; };
 
-        nixosSystem = { modules, ... } @ args:
-          import ./nixos/lib/eval-config.nix (args // {
-            modules =
-              let
-                moduleDeclarationFile =
-                  let
-                    # Even though `modules` is a mandatory argument for `nixosSystem`, it doesn't
-                    # mean that the evaluator always keeps track of its position. If there
-                    # are too many levels of indirection, the position gets lost at some point.
-                    intermediatePos = builtins.unsafeGetAttrPos "modules" args;
-                  in
-                    if intermediatePos == null then null else intermediatePos.file;
-
-                # Add the invoking file as error message location for modules
-                # that don't have their own locations; presumably inline modules.
-                addModuleDeclarationFile =
-                  m: if moduleDeclarationFile == null then m else {
-                    _file = moduleDeclarationFile;
-                    imports = [ m ];
-                  };
-
-              in
-              map addModuleDeclarationFile modules ++ [
-                {
-                  system.nixos.versionSuffix =
-                    ".${final.substring 0 8 (self.lastModifiedDate or self.lastModified or "19700101")}.${self.shortRev or "dirty"}";
-                  system.nixos.revision = final.mkIf (self ? rev) self.rev;
-                }
-              ];
+        nixosSystem = args:
+          import ./nixos/lib/eval-config.nix ({ lib = final; } // args // {
```

Comment:
This line isn't uncontroversial.

I'm having trouble weighing the utility of "more stuff available" against the subjective "good architecture". That's because I can't assume that we continue to have two different `lib`s prominently available in one repo.
I like "good architecture" too, but I think the die has been cast the moment `nixosSystem` was added to `lib`.
With that consideration, the solution is to add it to `nixpkgs/lib`, along with the experimental `nixos` attr above, and turn the whole thing into a nicely consistent user experience where `lib` is just `lib`.
So unless we have a real reason to keep the split, I think this change is vaguely in the right direction, but not complete or does not go to the root of the problem.

Hence, my question is, do we envisage a practical reason to keep `nixpkgs/lib` decoupled? This would be some feasible future action where the decoupling is actually necessary.
The only action I can think of is extracting `lib` into a separate repo. Although I question the utility of this, as pkgs, lib and nixos benefit a lot from being monorepos, I can see a solution for this, which is to extract the non-coupled parts - most of lib. Only a small layer that adds nixos and perhaps pkgs attrs would have to be separated out before extracting lib into a separate repo. We can even do this preemptively by making pkgs and nixos import a wrapper in a separate file, similar to what the flake does, but not just for the flake. Equivalently, we can move the non-coupled attrs into a separate overlay in lib, so that those parts can be accessed independently.

Reply:
Regardless of how we decide to organise the lib, I don't see a reason not to pass the "most accurate `lib`" to `eval-config.nix` (in this case `final`). If `nixosSystem` was defined in `nixpkgs/lib` as you suggest, surely it would have something like `{ lib = lib; } // args`, right? Otherwise we just force the user to add the `lib` argument themselves as soon as they want to use their own `lib` extensions in a module. I don't see how this is good architecture.

I think the decoupling debate is important and deserves its own issue/discussion/RFC if that's what it takes, but should be orthogonal to this PR.

Reply:
Phrased differently, `eval-config.nix` has a safe default value for `lib`, but `nixosSystem` knows better what `lib` should be since it's in the process of defining it.

Reply:
> but `nixosSystem` knows better what `lib` should be

Whether that's actually better is a judgement that would depend on the outcome of the discussion.

Alternative explanations exists that make it not better: conflating `flake.lib` and `nixpkgs/lib` if they're considered separate concepts. They could be.

These requirements make my mind go in circles. It's awful and I need to stop.

I hope others can agree on what's actually good for users.

Reply:
@cole-h would you happen to have an opinion about this?

Reply:
I just noticed

```nix
        nixos = import ./nixos/lib { lib = final; };
```

and now I'm confused

Reply:
Uh, yeah, I guess I remembered what you want when I wrote it and didn't think about it again since then.

Zooming out from the `lib = final` bit, that `nixos` attr is an experimental interface for testing and RFC 22-related refactoring. See https://github.com/NixOS/nixpkgs/pull/148456